### ISSUE TITLE
What would it take, approximately, to add cider-jack-in support for Babashka's NREPL server?

### ISSUE THREAD

**Issue by seylerius** [clojure-emacs/cider#2947]
**Is your feature request related to a problem? Please describe.**

I very much love working in CIDER, and I also quite like Babashka, but there are a few tiny gaps remaining in the integration between the two. Key among them is that `cider-jack-in` does not support Babashka yet, according to the CIDER documentation.

**Describe the solution you'd like**

I'd like to patch a version of CIDER and/or Babashka that can handle `cider-jack-in` integration, preferably with auto-detection based on a file having a Babashka shebang line, or an appropriate project file. I'm quite willing to do the legwork myself is someone will point me at the key gaps preventing the integration from happening. My first guesses include needing a detectable project file, or something in that vein.

**Describe alternatives you've considered**

I've considered making a wrapped Babashka REPL launcher that calls `cider-connect-clj`, but this will necessarily be more clumsy and kludgy. Merging these features in would be ideal, in the long run.

**Comment by bbatsov** [comment]
We mostly need a way to identify a project as a babashka project, the rest is trivial. Some file marker would be the easiest option, but an alternative would be to just force the project type via `.dir-local.el`. //cc @borkdude

**Comment by borkdude** [comment]
I'm considering something like "babashka.edn" . See https://github.com/borkdude/babashka/issues/680. But the vast majority of babashka "projects" are just single file scripts, so a couple of heuristics like looking at the shebang probably is good as well. 

Why not have an explicit `cider-jack-in-babashka` command?

Here is an idea for a script that starts the nREPL server for you and write a .nrepl-port file, maybe that also helps:

https://book.babashka.org/#repl

**Comment by bbatsov** [comment]
> Why not have an explicit cider-jack-in-babashka command?

Yeah, this would be very simple to do. It can just run the server in the current folder. 

> But the vast majority of babashka "projects" are just single file scripts, so a couple of heuristics like looking at the shebang probably is good as well.

This is also an option, but it will assume that someone has opened a babashka script.

**Comment by SignSpice** [comment]
I'm also interested in this.

**Comment by onetom** [comment]
We are interested too!
Our team is still at an early stage of exploration.
We are at our 3rd attempt to introduce Babashka into our workflows.
While we don't have crystalized desires yet, I would like to further the conversation by introducing some considerations.

I think `inf-clojure` would be often enough for our team, if it would support auto-completion in some simple way at least. We "just" need a definition for the `completion` feature here: https://github.com/clojure-emacs/inf-clojure/blob/master/inf-clojure.el#L131 (I made some experiments but I'm at the beginning of my Elisp journey, so I couldn't make anything work yet :)

Regarding "babashka project" detection:
The heuristics of looking for a shebang line makes sense and +1 for the `.dir-locals.el`-based explicit configuration, BUT in our case we are expecting to mix babashka with tools-deps projects (replacing bash scripts - as intended - and `Makefile`s), so our concept of a project doesn't have just a singular "`clojure-repl-type`".

I quite like the `C-u M-x inf-clojure` experience, which allows me to pick a REPL type explicitly. It's a bit manual, but would be a good initial step toward a support.

I'm not sure it's a good idea to extend the API of CIDER with an explicit `cider-jack-in-babashka` command, because that would just lead to a proliferation of such functions for other REPL types.

**Comment by borkdude** [comment]
Similar problem in https://github.com/babashka/babashka/issues/733: how to detect if a file is a babashka script or not for LSP mode.

**Comment by kirill-gerasimenko-da** [comment]
That would be awesome to have!

**Comment by timotheosh** [comment]
Don't let this issue rest. It would be good to have this option. I am wondering if a similar thing could be done for Lumo scripts. If the two project types are too different, well then just keep this open for babashka.

**Comment by benjamin-asdf** [comment]
isn't this fixed?

**Comment by bbatsov** [comment]
@benjamin-asdf CIDER supports jacking-in with BB, but there's no dedicated command for this.

**Comment by kirill-gerasimenko-da** [comment]
This is what I'm using (stolen from borkdude :)) 

```
(defun cider-jack-in-babashka ()
  "Start an babashka nREPL server for the current project and connect to it."
  (interactive)
  (let* ((default-directory (project-root (project-current t)))
         (process-filter (lambda (proc string)
                           "Run cider-connect once babashka nrepl server is ready."
                           (when (string-match "Started nREPL server at .+:\\([0-9]+\\)" string)
                             (cider-connect-clj (list :host "localhost"
                                                      :port (match-string 1 string)
                                                      :project-dir default-directory)))
                           ;; Default behavior: write to process buffer
                           (internal-default-process-filter proc string))))
    (set-process-filter
       (start-file-process "babashka" "*babashka*" "bb" "nrepl-server" "0")
       process-filter)))
```

**Comment by bbatsov** [comment]
See also the discussion here - https://github.com/clojure-emacs/cider/issues/3179 Basically, I was thinking it's probably more prudent to add some way to force `cider-jack-in-clj` to ask you for the REPL to start instead of having dedicated commands for everything supported.

**Comment by bbatsov** [comment]
I consider this addressed by #3300 and the new `cider-jack-in-universal` command.

**Comment by vemv** [comment]
> I consider this addressed by https://github.com/clojure-emacs/cider/pull/3300 and the new cider-jack-in-universal command.
